### PR TITLE
Proposal: `mapContaining` and `setContaining` asymmetric matchers

### PR DESCRIPTION
**Pull request by eventlistener**
## Description
* **New feature** (non-breaking change which adds functionality):
  `mapContaining` and `setContaining` asymmetric matchers that can be used for deep equality comparison of two maps and sets, respectively.

  Usage example for `mapContaining`:
  ```javascript
  expect(
    new Map([
      ['foo', 'bar'],
      ['baz', [1, 2]],
    ])
  ).toEqual(jasmine.mapContaining(
    new Map([
      [
        jasmine.stringMatching(/^foo/),
        'bar'
      ],
      [
        'baz',
        jasmine.any(Array)
      ],
    ])
  ));
  ```

  Usage example for `setContaining`:
  ```javascript
  expect(
    new Set([
      'foo', [1, 2], 'bar',
    ])
  ).toEqual(jasmine.setContaining(
    new Set([
      jasmine.any(Array), jasmine.stringMatching(/^foo/),
    ])
  ));
  ```

* **Bug fix** (non-breaking change which fixes an issue):
  The way `matchersUtil.contains` currently checks whether the `haystack` argument is a set does not work as intended in IE11. As a result, the matcher always returns `false` for sets. There doesn't seem to be an open issue in the tracker.

## Motivation and Context
The PR extends Jasmine's support of ES6 data structures. Even though the existing matcher `arrayContaining` can be used for such comparisons (if you convert the actual map/set to an array first), it's not as concise and convenient for this purpose.

## How Has This Been Tested?
Has been tested in Node via `npm test` and all the supported browsers (except Safari which I don't have) via `npm run serve`.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/core/matchers/matchersUtil.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ getJasmineRequireObj().matchersUtil = function(j$) {
     contains: function(haystack, needle, customTesters) {
       customTesters = customTesters || [];
 
-      if ((Object.prototype.toString.apply(haystack) === '[object Set]')) {
```
Comment (eventlistener) [inline src/core/matchers/matchersUtil.js]: IE11 returns `[object Object]`.

File: spec/core/matchers/matchersUtilSpec.js [hunk 2]
```diff
@@ -613,6 +639,26 @@ describe("matchersUtil", function() {
...
+
+    // documenting current behavior
+    it("fails (!) for objects that equal to a set member", function() {
+      jasmine.getEnv().requireFunctioningSets();
+
+      var set = new Set();
+      set.add({'foo': 'bar'});
+
+      expect(jasmineUnderTest.matchersUtil.contains(set, {'foo': 'bar'})).toBe(false);
+    });
```
Comment (eventlistener) [inline spec/core/matchers/matchersUtilSpec.js]: This is kind of confusing considering arrays are treated differently:
```javascript
expect([{'foo': 'bar'}]).toContain({'foo': 'bar'});    // passes
```
vs
```javascript
expect(new Set([{'foo': 'bar'}])).toContain({'foo': 'bar'});    // fails
```

**Comment by sgravrock** [comment]
Looks nice overall. I do see one problem: The new test files you added aren't actually running because their names don't end in `Spec.js`. If you can rename the spec files and double-check that they're doing everything you expect them to do then I think this will be ready to merge.

File: spec/core/matchers/matchersUtilSpec.js (hunk 2 above)
Comment (sgravrock) [reply spec/core/matchers/matchersUtilSpec.js]: Agreed, and it would be nice to change that but it'll have to go with 4.0.

**Comment by eventlistener** [comment]
The tests were certainly running in Node (they helped me find a bug in my code). I see now they were not running in browsers. I renamed the files, thanks!

**Comment by eventlistener** [comment]
Let me know if you want me to squash the commits.

**Comment by sgravrock** [comment]
> Let me know if you want me to squash the commits.

Yes please.

**Comment by eventlistener** [comment]
Done!

**Pull request merged**